pkg/controller/persistentvolume/types.go

```diff
@@ -115,19 +102,34 @@ func (pvIndex *persistentVolumeOrderedIndex) find(searchPV *api.PersistentVolume
 		// return the exact pre-binding match, if found
 		unboundVolumes := []*api.PersistentVolume{}
 		for _, volume := range volumes {
+			// volume isn't currently bound or pre-bound.
 			if volume.Spec.ClaimRef == nil {
-				// volume isn't currently bound or pre-bound.
 				unboundVolumes = append(unboundVolumes, volume)
 				continue
 			}
 
-			boundClaim := fmt.Sprintf("%s/%s", volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name)
-			if boundClaim == preboundClaim {
+			if claim.Name == volume.Spec.ClaimRef.Name && claim.Namespace == volume.Spec.ClaimRef.Namespace {
 				// exact match! No search required.
 				return volume, nil
 			}
 		}
 
+		// a claim requesting provisioning will have an exact match pre-bound to the claim.
+		// no need to search through unbound volumes.  The matching volume will be created by the provisioner
+		// and will match above when the claim is re-processed by the binder.
+		if keyExists(qosProvisioningKey, claim.Annotations) {
+			return nil, nil
+		}
```

Comment:
@saad-ali The binder will ignore a claim with the provisioningKey if it's not an exact match with a volume (i.e, the volume has a ClaimRef of the claim).  This might be non-obvious in the controller (though I have comments there saying so), but this block of code is where it is enforced.  Claims only bind to volumes provisioned for it, if the annotation is present.